cpp/src/arrow/util/iterator_test.cc

```diff
@@ -214,6 +255,270 @@ TEST(TestVectorIterator, RangeForLoop) {
+  auto collected_future = CollectAsyncGenerator(transformed);
+  ASSERT_TRUE(collected_future.Wait(5));
+  if (collected_future.is_finished()) {
+    ASSERT_OK_AND_ASSIGN(auto collected, collected_future.result());
+    ASSERT_EQ(0, collected.size());
+  }
+}
+
+TEST(TestAsyncUtil, Visit) {
+  auto generator = AsyncVectorIt({1, 2, 3});
+  unsigned int sum = 0;
+  auto sum_future = VisitAsyncGenerator<TestInt>(generator, [&sum](TestInt item) {
+    sum += item.value;
+    return Status::OK();
+  });
+  // Should be superfluous
+  sum_future.Wait();
+  ASSERT_EQ(6, sum);
+}
+
+TEST(TestAsyncUtil, Collect) {
+  std::vector<TestInt> expected = {1, 2, 3};
+  auto generator = AsyncVectorIt(expected);
+  auto collected = CollectAsyncGenerator(generator);
+  ASSERT_EQ(expected, *collected.result());
+}
+
+template <typename T>
+Transformer<T, T> MakeRepeatN(int repeat_count) {
+  int current_repeat = 0;
+  return [repeat_count, current_repeat](T next) mutable -> Result<TransformFlow<T>> {
+    current_repeat++;
+    bool ready_for_next = false;
+    if (current_repeat == repeat_count) {
+      current_repeat = 0;
+      ready_for_next = true;
+    }
+    return TransformYield(next, ready_for_next);
+  };
+}
+
+TEST(TestIteratorTransform, Repeating) {
+  auto original = VectorIt({1, 2, 3});
+  auto repeated = MakeTransformedIterator<TestInt, TestInt>(std::move(original),
+                                                            MakeRepeatN<TestInt>(2));
+  AssertIteratorMatch({1, 1, 2, 2, 3, 3}, std::move(repeated));
+}
+
+template <typename T>
+Transformer<T, T> MakeFilter(std::function<bool(T&)> filter) {
+  return [filter](T next) -> Result<TransformFlow<T>> {
+    if (filter(next)) {
+      return TransformYield(next);
+    } else {
+      return TransformSkip();
+    }
+  };
+}
+
+template <typename T>
+Transformer<T, T> MakeAbortOnSecond() {
+  int counter = 0;
+  return [counter](T next) mutable -> Result<TransformFlow<T>> {
+    if (counter++ == 1) {
+      return Status::Invalid("X");
+    }
+    return TransformYield(next);
+  };
+}
+
+TEST(TestIteratorTransform, SkipSome) {
+  // Exercises TransformSkip
+  auto original = VectorIt({1, 2, 3});
+  auto filter = MakeFilter<TestInt>([](TestInt& t) { return t.value != 2; });
+  auto filtered = MakeTransformedIterator(std::move(original), filter);
+  AssertIteratorMatch({1, 3}, std::move(filtered));
+}
+
+TEST(TestIteratorTransform, SkipAll) {
+  // Exercises TransformSkip
+  auto original = VectorIt({1, 2, 3});
+  auto filter = MakeFilter<TestInt>([](TestInt& t) { return false; });
+  auto filtered = MakeTransformedIterator(std::move(original), filter);
+  AssertIteratorMatch({}, std::move(filtered));
+}
+
+TEST(TestIteratorTransform, Abort) {
+  auto original = VectorIt({1, 2, 3});
+  auto transformed =
+      MakeTransformedIterator(std::move(original), MakeAbortOnSecond<TestInt>());
+  ASSERT_OK(transformed.Next());
+  ASSERT_RAISES(Invalid, transformed.Next());
+}
```

Comment:
For the `Transformer` no.  A `Transformer` cannot be called in parallel (this may be a limitation we have to revisit).   For an `AsyncGenerator` in general, perhaps, it is undefined.  Imagine your source is a CSV file stored on AWS and so you are issuing 8 reads in parallel.  It's entirely feasible that the 3rd read fails while the other reads succeed.

Reply:
Ok. Regardless of the intent, though, it's better to test it :-)

Reply:
I now test this case.